service/project/FileState.kt

```diff
@@ -188,7 +188,7 @@ class FileState internal constructor(
     }
 
     override fun mayOpenAndRun(content: String) {
-        if (!isRunnable || (!isOpen && !tryOpen())) return
```

Comment:
This is slightly awkward, because intuitively `isRunnable` really should be a gate for `mayOpenAndRun`. If the file isn't runnable, why are we allowed to run it?

Moreover, in the case where we are allowed to run our non-runnable file, why may we open it? The only way to run a text selection is by having the file open and selecting some text.

These awkwardnesses lead me to believe we should split the method up in two: `mayOpenAndRun()` and `mayRunSnippet(snippet: String)`. They'll behave similarly, but one of them doesn't need to check `isRunnable`.

Reply:
Addressed this - we've refactored mayOpenAndRun (at the interface and implementation level) to not take `content: String` as we already have `runContent` as a property of the interface.